compose/ui/ui-test-junit4/src/desktopMain/kotlin/androidx/compose/ui/test/junit4/matchers/MSSIMMatcher.kt

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright 2020 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.ui.test.junit4.matchers
+
+import androidx.annotation.FloatRange
+import kotlin.math.pow
+import org.jetbrains.skia.Color
+
+// Adopted copy of test/screenshot/screenshot/src/main/java/androidx/test/screenshot/matchers/MSSIMMatcher.kt
+
+/**
+ * Image comparison using Structural Similarity Index, developed by Wang, Bovik, Sheikh, and
+ * Simoncelli. Details can be read in their paper:
+ * https://ece.uwaterloo.ca/~z70wang/publications/ssim.pdf
+ */
+internal class MSSIMMatcher(@FloatRange(from = 0.0, to = 1.0) private val threshold: Double) :
+    BitmapMatcher {
+
+    companion object {
+        // These values were taken from the publication
+        private const val CONSTANT_L = 254.0
+        private const val CONSTANT_K1 = 0.00001
+        private const val CONSTANT_K2 = 0.00003
+        private val CONSTANT_C1 = (CONSTANT_L * CONSTANT_K1).pow(2.0)
+        private val CONSTANT_C2 = (CONSTANT_L * CONSTANT_K2).pow(2.0)
+        private const val WINDOW_SIZE = 10
+    }
+
+    override fun compareBitmaps(
+        expected: IntArray,
+        given: IntArray,
+        width: Int,
+        height: Int
+    ): MatchResult {
+        val SSIMTotal = calculateSSIM(expected, given, width, height)
+
+        val stats =
+            "[MSSIM] Required SSIM: $threshold, Actual " + "SSIM: " + "%.3f".format(SSIMTotal)
```

Comment:
Now we have this exception:
```
Image mismatch! Comparison stats: '[MSSIM] Required SSIM: 0.99, Actual SSIM: 0.976'
```

Before the change we had:
```
java.lang.AssertionError: java.lang.AssertionError: Image mismatch! Expected image compose/ui/ui-desktop/platform/androidx_compose_ui_platform_GraphicsLayerTest_rotationXYZ.png, actual: compose/ui/ui-desktop/platform/androidx_compose_ui_platform_GraphicsLayerTest_rotationXYZ_actual.png. FS location: /home/teamcity/agent/work/8a20760945d0aeba/golden
```

We need to add paths to files